2nd implementation of the Config Trait

TyOverby opened the pull request:
closes #160 
closes #156 
closes #164

Basic gist of it is that you can write the following 

```rust
use bincode::{serialize, config};

serialize(&vec![1,2,3]);
// or 
config().big_endian().limit(12).serialize(&vec![1,2,3]);
```

depending on if you want to configure bincode.

Adding new configuration options should be backwards compatible.  All of this should be 0 cost (assuming that the inliner does its work correctly)

TyOverby commented:
@dtolnay: this is a rewrite of my older PR with some suggestions from you taken into account

dtolnay commented on src/config.rs:
```diff
@@ -0,0 +1,266 @@
+    type Limit: SizeLimit + 'static;
+    type Endian: ByteOrder + 'static;
+
+    #[inline(always)]
```
I think the inline attribute only means anything on functions that have a body. I don't think this automatically applies to all implementations of the trait (but I could be wrong).

TyOverby replied:
Done.

TyOverby replied:
No effect.

dtolnay replied:
Let's remove it then. It is still present in the most recent revision.

TyOverby replied:
Oh yeah, I accidentally removed them from the default methods on OptionsExt instead.  I wonder if they are used on default methods?

I'll look into this a bit more.

TyOverby commented on src/config.rs:
```diff
@@ -0,0 +1,266 @@
+
+    /// TODO: Document
+    #[inline(always)]
+    pub fn with_no_limit(self) -> Config {
```
Done.

TyOverby commented on src/config.rs:
```diff
@@ -0,0 +1,266 @@
+
+    /// TODO: Document
+    #[inline(always)]
+    pub fn with_limit(self, limit: u64) -> Config {
```
Done.

TyOverby commented:
@dtolnay: any more feedback on the design?  

If not, I'm going to fill out the doc comments and merge.

TyOverby commented:
Oh, one issue is that I'm no longer exporting BincodeRead, so people that relied on that (servo) will break.

I wonder if I could incorperate their custom BincodeRead into bincode and provide an unsafe fn for using it.

dtolnay commented on src/config.rs:
```diff
@@ -0,0 +1,251 @@
+
+    /// TODO: Document
+    #[inline(always)]
+    pub fn serialized_size_bounded<T: ?Sized + serde::Serialize>(&self, t: &T, bound: u64) -> Option<u64> {
```
I haven't thought through which way is better, but did you consider dropping this method and changing `serialized_size` to return a Result and then recommending `bincode::config().limit(bound).serialized_size(&t)`?

TyOverby replied:
That makes way too much sense.

dtolnay commented on src/config.rs:
```diff
@@ -0,0 +1,251 @@
+
+    /// TODO: Document
+    #[inline(always)]
+    pub fn serialized_size<T: ?Sized + serde::Serialize>(&self, t: &T) -> u64 {
```
I think this signature with u64 instead of Result is fine but the documentation definitely needs to explain what happens if serialization fails.

On the other hand there is some value in exposing that serialization failed in `serialized_size` before the caller goes on to invoke `serialize_into` which may allocate a bunch of memory and do a bunch of work. I left a comment on this below.

TyOverby replied:
Nah, I think that Result<u64> is a pretty good move.

dtolnay commented on src/internal.rs:
```diff
@@ -138,51 +17,45 @@ impl serde::ser::Error for Error {
         }
         None => {
-            let size = serialized_size(value) as usize;
+            let size = serialized_size(value, &mut options) as usize;
```
Whether or not we change to Result in the public `serialized_size`, I think this should return early if `serialized_size` failed.

TyOverby replied:
Agreed.  This has been changed to short circuit

TyOverby commented:
Pretty much done with this PR.  I added the `deserialize_from_custom` method, and updated the docs pretty substantially.

I'm going to talk with servo to see if this meets their needs and then I'll hit merge.

TyOverby commented:
I talked to @Gankro from servo and I'm pretty sure that with the `with_deserializer` method, we'll be able to hit all their requirements.
